[loader] Clean up locking around adding assemblies to domain/ALC

CoffeeFlux opened the pull request:
This cleans up three components of assembly loading that could potentially cause races/crashes.

First, both `add_assemblies_to_domain` and `add_assembly_to_alc` now ignore resolved references on netcore. We were seeing a crash on CI related to an assertion in here that resulted from two threads using the same MonoImage simultaneously as one is being loaded in with LoadFile, and this should solve that. Fixes #2305 

Second, we take both the domain and ALC loaded assemblies locks simultaneously around the assembly being added to both lists, ensuring they stay in sync.

Finally, MonoImage now takes the image lock when writing to references instead of using the global `assemblies_mutex` in assembly.c for all images. This potentially helps reduce contention. Work here is not complete, but for this PR it's good enough. See #32889

CoffeeFlux committed:
[loader] Clean up ALC/domain assembly cache locking and lift assertion

CoffeeFlux commented:
This is failing in `add_assemblies_to_domain` with the new assert, but it's failing much more frequently than the ALC one, which is good! Should be easier to track down how that reference is being resolved.

CoffeeFlux committed:
Add image references_lock to avoid one global lock for all images

There are probably more places we could take it, but we have to be careful since it's an OS lock (by need, I think, thanks to the embedding API).

CoffeeFlux committed:
Fix infinite recursion on legacy

CoffeeFlux commented on src/mono/mono/metadata/appdomain.c:
```diff
@@ -128,7 +128,7 @@ static gboolean
 
 static void
-add_assemblies_to_domain (MonoDomain *domain, MonoAssembly *ass, GHashTable *hash);
+add_assemblies_to_domain (MonoDomain *domain, MonoAssembly *ass, GHashTable *ht);
```
Maybe it's not worth the churn; I just noticed they were inconsistent and thought I'd fix it.

lambdageek commented on src/mono/mono/metadata/metadata-internals.h:
```diff
@@ -431,6 +431,7 @@ struct _MonoImage {
 	 */
 	MonoAssembly **references;
 	int nreferences;
+	mono_mutex_t references_lock;
```
I wonder what it would take to split up `MonoImage` into a base class (say `MonoEcmaImage`) that only knows about tables and such and make the current `MonoImage` a subclass for the main image of an assembly.  And we could have a separate `MonoPdbImage` or `MonoDeltaImage` that only has a subset of the concrete fields that are relevant for PPDBs or E&C images.

vargaz commented on src/mono/mono/metadata/assembly.c:
```diff
@@ -1783,15 +1793,15 @@ mono_assembly_load_reference (MonoImage *image, int index)
 	 * it inside a critical section.
 	 */
-	mono_assemblies_lock ();
+	mono_image_references_lock (image);
```
There is already a mono_image_lock () low level lock.

lambdageek replied:
@vargaz what are the reasons to prefer using the image lock instead of having a separate lock just for this field?

CoffeeFlux replied:
@vargaz ping

CoffeeFlux commented:
Waiting to hear from Zoltan.

CoffeeFlux committed:
Use image lock for references instead of its own lock

Contention is hopefully minimal, but switch back to a dedicated lock if it's a problem
